### ISSUE TITLE
Access Denied When Using Existing IDS4

### ISSUE THREAD

**Issue by CShelton11** [skoruba/IdentityServer4.Admin#276]
Have an existing IDS4 server.
Trying to login to admin using the uxisting ids4.
Login is successful and redirect occurs.
Getting Access denied error.

Example token seen below:

{
   "nbf":1557793643,
   "exp":1557793943,
   "iss":"https://localhost:44375",
   "aud":"mvc",
   "nonce":"636933904431764060.MTI0NzUzOWEtM2QxYi00NzQ1LTg4M2YtOGU1ZWM5ODdmZmNiYjk5OTMyZGUtOWE0Zi00YzFhLTkzYzctM2I4OTYxZTI2YTNl",
   "iat":1557793643,
   "at_hash":"zBDLqjCsFm0ubnpGwdcliw",
   "sid":"d4bdbaa6554e0db1908c65042baf3f14",
   "sub":"f86e073b-4500-476d-8e05-b77a95cfca83",
   "auth_time":1557793637,
   "idp":"local",
   "name":"Chris lastname",
   "given_name":"Chris",
   "family_name":"lastname",
   "email":"clastname@company.com",
   "roles":"['Admin']",
   "preferred_username":"Clastname",
   "amr":[
      "pwd"
   ]
}

I have the identity server configured to store the role information as ['admin'] in the database.
Am I storing the information incorrectly in the database, using the wrong role name, or something similar?

I have the admin const AdministrationRole = "Admin";

What else could cause the issue?

**Comment by CShelton11** [comment]
So curious as to exactly how this is supposed to work.
In my ids 4 configuration I added the following:

1. Scope (Example)
Id	Scope	ClientId
9	roles	2

2. Resource (Example)
Id	Enabled	Name	DisplayName	Description	Required	Emphasize	ShowInDiscoveryDocument	Created	Updated	NonEditable
4	1	roles	User roles	Your role information	0	1	1	2019-05-01 00:00:00.0000000	NULL	0

3. Identity Resource (Example)
Id	Type	IdentityResourceId
17	roles	4

4. AspNetUserClaim (Example)
Id	UserId	ClaimType	ClaimValue
25	f86e073b-4500-476d-8e05-b77a95cfca83	roles	['Admin']

I have also tried going into the AspNetRoles and Adding a role there (Example):
Id	Name	NormalizedName	ConcurrencyStamp
1	Admin	ADMIN	537213F0-1FE8-432C-8CE4-09CA83D47C2E

I also associated the user to that role in AspNetUserRoles (Example)
UserId	RoleId
f86e073b-4500-476d-8e05-b77a95cfca83	1

What am I missing here?

**Comment by CShelton11** [comment]
Please let me know if you need additional info here as I am stuck

**Comment by skoruba** [comment]
Please take a look at configuration of client and resources here:
https://github.com/skoruba/IdentityServer4.Admin/tree/master/src/Skoruba.IdentityServer4.Admin/Configuration/IdentityServer

And in StartupHelpers.cs - you can check setup of middleware for OpenIdConnect auth.

You can check User’s claims in MVC app - if roles are correctly stored in cookie.

Thanks

**Comment by CShelton11** [comment]
Thanks.  Based on what you've sent I believe the issue is in new IdentityResource("roles", "Roles", new[] { "role" })

Since my implementation is storing the resources in the database I am just trying to manually enter the data.  I'm guess that there is something different about the data stored for that particular identity resource.  I have nothing in my tables that indicated the info should be stored in the AspNetUserClaims table with key of role, nor is there anything that indicates that information will be an array of role.  In my circumstance, I simply have a single entry in that table titled roles with Admin as the value.  

Just need to determine what that difference is I guess.  If you have an example of entries in the database that match those expected that would be helpful.  Otherwise thanks for the pointer...

**Comment by CShelton11** [comment]
Found the issue.  I had included the Resource with name "roles" which was correct, but the identity claims table had 'roles' in the types field as opposed to 'role'.  Silly mistake.  The unintended consequence of having the resource name with the exact same name of the identity claim cause the object to be serialized as a string and resulted in admin being unable to read the token properly.   Thanks for the help on this.
